### PR TITLE
Tests for MergeWithSCMExtention.java 

### PR DESCRIPTION
**Pull request by Loghijiaha**
## Tests for MergeWithSCMExtention.java 

The pull request is for the merging test that testing the src/main/java/jenkins/plugin/git/MergeWithGitSCMExtension.java. 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] Dependency or infrastructure update

### CODE REVIEW

File: src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    private TestGitRepo repo;
+    private String baseName;
+    private String baseHash;
+    private String MASTER_FILE = "commitFileBase";
+
+    public void before() throws Exception {
+        repo = new TestGitRepo("repo", tmp.newFolder(), listener);
+        // make an initial commit to master and get hash
+        this.baseHash=repo.commit(MASTER_FILE, repo.johnDoe, "Initial Commit");
```
Comment (res0nance) [inline src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java]:
```suggestion
        this.baseHash = repo.commit(MASTER_FILE, repo.johnDoe, "Initial Commit");
```

File: src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        // set the base name as HEAD
+        this.baseName=Constants.MASTER;
```
Comment (res0nance) [inline src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java]:
```suggestion
        this.baseName = Constants.MASTER;
```

File: src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        assertEquals(firstBuild.getNumber(), buildData.lastBuild.getBuildNumber());
+        Revision firstMarked = buildData.lastBuild.getMarked();
+        Revision firstRevision = buildData.lastBuild.getRevision();
+        assertNotNull(firstMarked);
+        assertNotNull(firstRevision);
+
+        // delete integration branch successfully and commit successfully
+        repo.git.deleteBranch("integration");
+        repo.git.checkoutBranch("integration", "master");
+        this.baseName=Constants.HEAD;
```
Comment (res0nance) [inline src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java]:
```suggestion
        this.baseName = Constants.HEAD;
```

File: src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        assertEquals(firstMarked, gitSCM.getBuildData(firstBuild).lastBuild.getMarked());
+        assertEquals(firstRevision, gitSCM.getBuildData(firstBuild).lastBuild.getRevision());
+    }
+
+    @Override
+    protected GitSCMExtension getExtension() {
+        return new MergeWithGitSCMExtension(baseName,baseHash);
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (Loghijiaha) [reply src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java]: @MarkEWaite No. They are not necessary. I will remove it.

File: src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import org.jenkinsci.plugins.gitclient.MergeCommand;
+import org.junit.Test;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.Assert.*;
+import static org.junit.Assert.assertEquals;
```
Comment (MarkEWaite) [inline src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java]: Having imported all Assert classes with the wildcard import, we don't need to import assertEquals explicitly. I prefer the wildcard import of the Assert classes, so this line can be removed.

The [SCM coding style guide](https://github.com/jenkinsci/scm-api-plugin/blob/master/CONTRIBUTING.md#imports) allows wildcard imports of JUnit and HJamcrest classes.
```suggestion
```

File: src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import hudson.plugins.git.extensions.impl.LocalBranch;
+import hudson.plugins.git.extensions.impl.PreBuildMerge;
+import hudson.plugins.git.util.BuildData;
+import junit.framework.TestCase;
+import nl.jqno.equalsverifier.EqualsVerifier;
+import org.eclipse.jgit.lib.Constants;
+import org.jenkinsci.plugins.gitclient.MergeCommand;
+import org.junit.Test;
+
+import java.io.*;
```
Comment (MarkEWaite) [inline src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java]: The [CONTRIBUTING guide](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc#contributing-to-the-git-plugin) states that new code should follow the [SCM plugin coding style](https://github.com/jenkinsci/scm-api-plugin/blob/master/CONTRIBUTING.md#code-style-guidelines).

Those guidelines say that [wildcard imports are banned](https://github.com/jenkinsci/scm-api-plugin/blob/master/CONTRIBUTING.md#imports).  If you'd be willing to replace this wildcard import with regular imports, that would help reduce the style differences between this and other new code being added to the plugin.

File: src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (MarkEWaite) [reply src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java]: Interesting.  The unexpected characters are still there in my view on GitHub.  I'll attempt a proposed change to remove them.

File: src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java (hunk 6 above)
Comment (Loghijiaha) [reply src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java]: Yes. I forgot to check coding style. Now,  I optimized and removed the banned import style.

Comment (MarkEWaite) [reply src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java]: Thanks so much! That is kind of you to make the update when the existing code is painfully inconsistent in its formatting.

File: src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java [hunk 7]
```diff
@@ -92,4 +92,4 @@ protected GitSCMExtension getExtension() {
         return new MergeWithGitSCMExtension(baseName,baseHash);
     }
 
-}
+}
```
Comment (darxriggs) [inline src/test/java/jenkins/plugins/git/MergeWithGitSCMExtensionTest.java]: There should always be a newline at the end of a  file. GitHub also shows the absence of it as two red symbols. If you hover of it, it says "No newline at end of file". That's exactly what git shows on the command line when a `git show --patch <commit id>` is performed.

**Pull request merged**